commons-core/src/test/scala/com/avsystem/commons/serialization/json/JsonStringInputOutputTest.scala

```diff
@@ -0,0 +1,206 @@
+    val lo = output.writeList()
+    jsons.foreach(json => lo.writeElement().writeRawJson(json))
+    lo.finish()
+    val jsonList = sb.toString
+
+    assert(jsonList == jsons.mkString("[", ",", "]"))
+
+    val input = new JsonStringInput(new JsonReader(jsonList))
+    val li = input.readList()
+    val resBuilder = new ListBuffer[String]
+    while (li.hasNext) {
+      resBuilder += li.nextElement().readRawJson()
+    }
+    assert(resBuilder.result() == jsons)
+  }
+
+  test("raw json object test") {
+    val jsons = IListMap("a" -> "123", "b" -> "null", "c" -> "\"str\"", "d" -> "4.5", "e" -> "[1,2,3]", "f" -> "{\"a\": 123, \"b\": 3.14}")
+    val sb = new JStringBuilder
+    val output = new JsonStringOutput(sb)
+    val oo = output.writeObject()
+    jsons.foreach { case (fn, fv) => oo.writeField(fn).writeRawJson(fv) }
+    oo.finish()
+    val jsonList = sb.toString
+
+    assert(jsonList == jsons.map({ case (k, v) => s""""$k":$v""" }).mkString("{", ",", "}"))
+
+    val input = new JsonStringInput(new JsonReader(jsonList))
+    val oi = input.readObject()
+    val resBuilder = IListMap.newBuilder[String, String]
+    while (oi.hasNext) {
+      val fi = oi.nextField()
+      resBuilder += ((fi.fieldName, fi.readRawJson()))
+    }
+    assert(resBuilder.result() == jsons)
+  }
+
+  def roundtrip[T: GenCodec](name: String)(values: Seq[T]): Unit = {
+    test(name) {
+      val serialized = values.map(write[T])
+      val deserialized = serialized.map(read[T])
+      deserialized shouldBe values
+    }
+  }
+
+  roundtrip("integers")(Seq(Int.MinValue, -42, 0, 42, Int.MaxValue))
+
+  roundtrip("longs")(Seq(Int.MaxValue.toLong + 1, Long.MinValue, -783868188, 0, 783868188, Long.MaxValue))
+
+  roundtrip("doubles")(Seq(
+    Double.MinPositiveValue, Double.MinValue, -1.750470182E9, Double.MaxValue, Double.PositiveInfinity, Double.NegativeInfinity)
+  )
+  roundtrip("booleans")(Seq(false, true))
+
+  roundtrip("strings")(Seq("", "a።bc\u0676ąቢść➔Ĳ"))
+
+  roundtrip("simple case classes")(Seq(TestCC(5, 123L, 432, true, "bla", 'a' :: 'b' :: Nil)))
+
+  test("NaN") {
+    val value = Double.NaN
+
+    val serialized = write(value)
+    val deserialized = read[Double](serialized)
+
+    assert(java.lang.Double.isNaN(deserialized))
+  }
+
+  test("scientific") {
+    val value = -1.750470182E9
+    val test = value.toString
+    val serialized = Seq("-1.750470182E+9", "-1.750470182E9", test)
+    val deserialized = serialized.map(read[Double])
+
+    deserialized should contain only value
+  }
+
+  test("serialize and deserialize nested case classes") {
+    val test: TestCC = TestCC(5, 123L, 432, true, "bla", 'a' :: 'b' :: Nil)
+    val test2: TestCC = TestCC(-35, 1L, 432, true, "blsddf sdg  \"{,}[,]\"a", 'a' :: 'b' :: Nil)
+    val nested: NestedTestCC = NestedTestCC(-123, test, test2)
+    val serialized = write(nested)
+    val deserialized = read[NestedTestCC](serialized)
+
+    deserialized should be(nested)
```

Comment:
`<rant>`
This is also what annoys me about this syntax. Are we using infix/postfix or parens? Here we have a weird mix of both.
`</rant>`

Reply:
Maybe I'm just too used to how it works. I'll use `shouldBe` instead.